### ISSUE TITLE
1.8.0 launchd issue on mac os x

### ISSUE THREAD

**Issue by shauvik** [facebook/osquery#2262]
I had osquery 1.7.4 running well on my machine but when I installed 1.8.0 over it to upgrade.

However, `osqueryctl status` reports:
```
$ sudo osqueryctl status
com.facebook.osqueryd is not running

$ sudo osqueryctl start
/Library/LaunchDaemons/com.facebook.osqueryd.plist: service already loaded
```

Unloading the launch daemon manually and restarting didn't fix.

I was able to fix this by uninstalling osquery completely and installing 1.7.4 back:
```
$ sudo rm -rf /private/var/log/osquery /private/var/osquery /usr/local/bin/osquery* \
    /Library/LaunchDaemons/com.facebook.osquery.plist
$ sudo pkgutil --forget com.facebook.osquery
$ sudo installer -pkg osquery-1.7.4.pkg -target /
```

**Comment by shauvik** [comment]
osquery-1.7.5 also works but only the latest one seems broken.

**Comment by theopolis** [comment]
Interesting, where did you get the `pkg` from? And can I see the content of `/Library/LaunchDaemons/com.facebook.osquery.plist`?

Here's my short debugging session:
```
/tmp  » sudo installer -pkg ./osquery-1.8.0.pkg -target /
installer: Package name is osquery-1.8.0
installer: Installing at base path /
installer: The install was successful.
/tmp  » sudo osqueryctl status                           
com.facebook.osqueryd is running. pid: 80389
/tmp  » sudo osqueryctl start  
/Library/LaunchDaemons/com.facebook.osqueryd.plist: service already loaded
/tmp  » sudo osqueryctl stop 
/tmp  » sudo osqueryctl status
com.facebook.osqueryd is not running
/tmp  » sudo osqueryctl start 
/tmp  » sudo osqueryctl status
com.facebook.osqueryd is running. pid: 80515
/tmp  » osqueryi --version
osqueryi version 1.7.7-30-g7f304a0
/tmp  » cat /Library/LaunchDaemons/com.facebook.osqueryd.plist 
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
  <key>KeepAlive</key>
  <true/>
  <key>Disabled</key>
  <false/>
  <key>Label</key>
  <string>com.facebook.osqueryd</string>
  <key>ProgramArguments</key>
  <array>
    <string>/usr/local/bin/osqueryd</string>
    <string>--flagfile=/private/var/osquery/osquery.flags</string>
  </array>
  <key>RunAtLoad</key>
  <true/>
  <key>ThrottleInterval</key>
  <integer>60</integer>
</dict>
</plist>
/tmp  » 
```

**Comment by shauvik** [comment]
Note that your `osqueryi --version` reports a wrong version. Not sure if you're somehow using an older version as 1.8.0.

Here is the working launchd config for 1.7.5
```
$ osqueryi -version
osqueryi version 1.7.5
~ $ cat /Library/LaunchDaemons/com.facebook.osqueryd.plist
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
  <key>KeepAlive</key>
  <true/>
  <key>Disabled</key>
  <false/>
  <key>Label</key>
  <string>com.facebook.osqueryd</string>
  <key>ProgramArguments</key>
  <array>
    <string>/usr/local/bin/osqueryd</string>
    <string> --flagfile="/private/var/osquery/osquery.flags"</string>
  </array>
  <key>RunAtLoad</key>
  <true/>
  <key>ThrottleInterval</key>
  <integer>60</integer>
</dict>
</plist>
```


Here are my steps to install 1.8.0
```
$ sudo osqueryctl stop
Password: 
~ $ sudo rm -rf /private/var/log/osquery /private/var/osquery /usr/local/bin/osquery* \
>     /Library/LaunchDaemons/com.facebook.osquery.plist
~ $ sudo pkgutil --forget com.facebook.osquery
Forgot package 'com.facebook.osquery' on '/'.
~ $
~ $ sudo installer -pkg osquery-1.8.0.pkg -target /
installer: Package name is osquery-1.8.0
installer: Installing at base path /
installer: The install was successful.
~ $ sudo osqueryctl status   
com.facebook.osqueryd is not running
~ $ sudo osqueryctl start
No osquery config file found at /var/osquery/osquery.conf
See '/var/osquery/osquery.example.conf' for an example config.
~ $ sudo cp /var/osquery/osquery.example.conf /var/osquery/osquery.conf
~ $ sudo osqueryctl start
~ $ sudo osqueryctl stop
~ $ sudo osqueryctl start
~ $ cat /Library/LaunchDaemons/com.facebook.osqueryd.plist
cat: /Library/LaunchDaemons/com.facebook.osqueryd.plist: No such file or directory
~ $ sudo cp /var/osquery/com.facebook.osqueryd.plist /Library/LaunchDaemons/
~ $ sudo osqueryctl start
~ $ sudo osqueryctl status
com.facebook.osqueryd is not running
~ $ cat /Library/LaunchDaemons/com.facebook.osqueryd.plist 
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
  <key>KeepAlive</key>
  <true/>
  <key>Disabled</key>
  <false/>
  <key>Label</key>
  <string>com.facebook.osqueryd</string>
  <key>ProgramArguments</key>
  <array>
    <string>/usr/local/bin/osqueryd</string>
    <string>--flagfile=/private/var/osquery/osquery.flags</string>
  </array>
  <key>RunAtLoad</key>
  <true/>
  <key>ThrottleInterval</key>
  <integer>60</integer>
</dict>
</plist>
```

**Comment by theopolis** [comment]
I'll need to dig deeper but my results are different:
```
/tmp  » sudo osqueryctl stop
/tmp  » rm -rf /Library/LaunchDaemons/com.facebook.osquery.plist
/tmp  » sudo pkgutil --forget com.facebook.osquery
Forgot package 'com.facebook.osquery' on '/'.
/tmp  » ls -la /Library/LaunchDaemons| grep osquery
/tmp  [1]»sudo installer -pkg ./osquery-1.8.0.pkg -target /
installer: Package name is osquery-1.8.0
installer: Installing at base path /
installer: The install was successful.
/tmp  » sudo osqueryctl status
com.facebook.osqueryd is not running
/tmp  » sudo osqueryctl start
/tmp  » sudo osqueryctl stop 
/tmp  » sudo osqueryctl start
/tmp  » cat /Library/LaunchDaemons/com.facebook.osqueryd.plist 
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
  <key>KeepAlive</key>
  <true/>
  <key>Disabled</key>
  <false/>
  <key>Label</key>
  <string>com.facebook.osqueryd</string>
  <key>ProgramArguments</key>
  <array>
    <string>/usr/local/bin/osqueryd</string>
    <string>--flagfile=/private/var/osquery/osquery.flags</string>
  </array>
  <key>RunAtLoad</key>
  <true/>
  <key>ThrottleInterval</key>
  <integer>60</integer>
</dict>
</plist>
/tmp  » sudo osqueryctl status                                
com.facebook.osqueryd is running. pid: 55787
```

**Comment by theopolis** [comment]
When you run `start`:
```bash
start() {
  check_config
  if [ $OS = "darwin" ]; then
    cp $PLIST_INSTALLATION_PATH $PLIST_PATH
    launchctl load $PLIST_PATH
  else
    exec_with_env "service osqueryd start"
  fi
}
```

Where:
```
  PLIST_DOMAIN="com.facebook.osqueryd"
  PLIST_PATH="/Library/LaunchDaemons/$PLIST_DOMAIN.plist"
  PLIST_INSTALLATION_PATH="/var/osquery/$PLIST_DOMAIN.plist"
```

**Comment by shauvik** [comment]
I tried it on another machine and got the same problem with 1.8.0

```
Shauviks-Air:~ shauvik$ sudo rm -rf /private/var/log/osquery /private/var/osquery /usr/local/bin/osquery* \
>     /Library/LaunchDaemons/com.facebook.osquery.plist
Password:
Sorry, try again.
Password:
Shauviks-Air:~ shauvik$ 
Shauviks-Air:~ shauvik$ 
Shauviks-Air:~ shauvik$ sudo pkgutil --forget com.facebook.osquery
No receipt for 'com.facebook.osquery' found at '/'.
Shauviks-Air:~ shauvik$ 
Shauviks-Air:~ shauvik$ curl -O https://osquery-packages.s3.amazonaws.com/darwin/osquery-1.8.0.pkg
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 6657k  100 6657k    0     0  1902k      0  0:00:03  0:00:03 --:--:-- 1902k
Shauviks-Air:~ shauvik$ osqueryi -version
-bash: osqueryi: command not found
Shauviks-Air:~ shauvik$ sudo installer -pkg osquery-1.8.0.pkg -target /
installer: Package name is osquery-1.8.0
installer: Installing at base path /
installer: The install was successful.
Shauviks-Air:~ shauvik$ sudo osqueryctl status   
com.facebook.osqueryd is not running
Shauviks-Air:~ shauvik$ 
Shauviks-Air:~ shauvik$ sudo cp /var/osquery/osquery.example.conf /var/osquery/osquery.conf
Shauviks-Air:~ shauvik$ 
Shauviks-Air:~ shauvik$ sudo osqueryctl start
Shauviks-Air:~ shauvik$ sudo osqueryctl status   
com.facebook.osqueryd is not running
Shauviks-Air:~ shauvik$ 
Shauviks-Air:~ shauvik$ cat /Library/LaunchDaemons/com.facebook.osqueryd.plist
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
  <key>KeepAlive</key>
  <true/>
  <key>Disabled</key>
  <false/>
  <key>Label</key>
  <string>com.facebook.osqueryd</string>
  <key>ProgramArguments</key>
  <array>
    <string>/usr/local/bin/osqueryd</string>
    <string>--flagfile=/private/var/osquery/osquery.flags</string>
  </array>
  <key>RunAtLoad</key>
  <true/>
  <key>ThrottleInterval</key>
  <integer>60</integer>
</dict>
</plist>
Shauviks-Air:~ shauvik$ sudo launchctl list  | grep osquery
-	1	com.facebook.osqueryd
Shauviks-Air:~ shauvik$ sudo osqueryctl status   
com.facebook.osqueryd is not running
Shauviks-Air:~ shauvik$ sudo launchctl unload com.facebook.osquery
/Users/shauvik/com.facebook.osquery: No such file or directory
Shauviks-Air:~ shauvik$ sudo launchctl unload /Library/LaunchDaemons/com.facebook.osqueryd.plist 
Shauviks-Air:~ shauvik$ sudo osqueryctl status   
com.facebook.osqueryd is not running
Shauviks-Air:~ shauvik$ sudo osqueryctl start
Shauviks-Air:~ shauvik$ sudo osqueryctl status   
com.facebook.osqueryd is not running
Shauviks-Air:~ shauvik$ sudo launchctl list  | grep osquery
-	1	com.facebook.osqueryd
Shauviks-Air:~ shauvik$ sudo launchctl unload /Library/LaunchDaemons/com.facebook.osqueryd.plist 
Shauviks-Air:~ shauvik$ sudo launchctl load /Library/LaunchDaemons/com.facebook.osqueryd.plist 
Shauviks-Air:~ shauvik$ sudo osqueryctl status   
com.facebook.osqueryd is not running
Shauviks-Air:~ shauvik$ sudo launchctl unload /Library/LaunchDaemons/com.facebook.osqueryd.plist 
Shauviks-Air:~ shauvik$ sudo /usr/local/bin/osqueryd --flagfile=/private/var/osquery/osquery.flags
/private/var/osquery/osquery.flags: No such file or directory
```

**Comment by texereda** [comment]
Hi there, I do confirm the same behaviour reported by @shauvik, I'm currently using the latest osquery version (1.8.2).

**Comment by theopolis** [comment]
Is anyone interested in submitting a fix? :+1:

**Comment by texereda** [comment]
Hi there, got a fix, please see my pull request (and the attached disclaimer) ;-)

**Comment by theopolis** [comment]
Interesting fix! I'm hesitant to merge just yet. I do not want to mess up any initial configuration requirements we have.

The expectation is that the daemon/LD should not start until the machine has a valid configuration or flagfile.

**Comment by texereda** [comment]
Fair enough but once the com.facebook.osqueryd.plist file is in the
PLIST_PATH, the launchd will try to restart the osquery daemon every 60
seconds, messing the system logs until the problem is fixed (valid config
file and valid/empty flagfile).

Maybe it's worth adding a check in osqueryctl start and make sure that the
osqueryd is running correctly, and if not maybe we should remove the plist
file from the PLIST_PATH.

It has to be noted that the launchd/launchctl ecosystem was refactored by
Apple in OS X 10.10 (for reference:
http://www.newosxbook.com/articles/8-10.10.html)  and the load and unload
commands are now being considered legacy, that's why on the "osqueryctl
stop" command under OS X 10.11.6  I always get an error 143 from launchd:

$ sudo osqueryctl stop
Aug  4 16:26:54 Sidera com.apple.xpc.launchd[1]
(com.facebook.osqueryd[8712]): Service exited with abnormal code: 143
$ sudo launchctl error 143
143: Requested entry was cached

so another check might be required in order to determine the exact OS
version and then use the appropriate non-legacy commands for
starting/stopping the osquery daemon.

Other projects have been impacted by this, see:
https://github.com/Homebrew/legacy-homebrew/issues/33259

Should I open another issue for this, just to keep it for the records ?

**Comment by theopolis** [comment]
Take a look at the PR I just opened, let me know if that fixes the odd-initial state checks.

**Comment by theopolis** [comment]
Ah the issue is much simpler. The `--flagfile` does not exist so the GFlags parsing is stopping the daemon. That contrasts with our expectation that you have configured either it or a `filesystem` config plugin's `--config_path` (default `/var/osquery/osquery.conf`) on OS X.

Either GFlag's expectation needs to be relaxed or the LaunchDaemon configuration needs to optionally supply the path.
